File: src/main/java/me/narutopig/bors/util/Enchanting.java
package me.narutopig.bors.util;

import org.bukkit.ChatColor;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static me.narutopig.bors.util.General.toRoman;

public class Enchanting {
    public static boolean hasEnchant(ItemStack item, Enchantment enchantment) {
        return item.getEnchantments().get(enchantment) != null;
    }

    public static boolean hasEnchant(Map<Enchantment, Integer> enchants, Enchantment enchantment) {
        return enchants.get(enchantment) != null;
    }

    public static boolean hasConflict(Enchantment enchantment, Map<Enchantment, Integer> enchants) {
        for (Enchantment e : enchants.keySet()) {
            if (enchantment.conflictsWith(e)) return true;
        }
        return false;
    }

    public static int calculateLevel(Enchantment e, int a, int b) {
        /*
        Calculates result of combining the two levels
         */
        int res;
        if (a != b) res = Math.max(a, b);
        else res = a + 1;
        return Math.min(res, e.getMaxLevel());
    }

    public static boolean canEnchant(ItemStack item, Enchantment enchantment) {
        return enchantment.canEnchantItem(item) && enchantment.getItemTarget().includes(item);
    }

    public static void addEnchant(ItemStack item, Enchantment enchantment, int level) {
        if (!canEnchant(item, enchantment)) return;
        ItemMeta itemMeta = item.getItemMeta();

        if (itemMeta == null) return;

        List<String> lore = itemMeta.getLore();

        if (lore == null) lore = new ArrayList<>();

        if (hasEnchant(item, enchantment)) {
            int i = 0;
            for (String line : lore) {
                if (ChatColor.stripColor(line).split(" ")[0].equals(enchantment.getName())) {
                    String newLine = ChatColor.GREEN + enchantment.getName() + " " + toRoman(level);
                    lore.set(i, newLine);
                    break;
                }
                i++;
            }
        } else {
            String newLine = ChatColor.GREEN + enchantment.getName() + " " + toRoman(level);
            lore.add(newLine);
        }

        itemMeta.setLore(lore);
        item.setItemMeta(itemMeta);

        item.removeEnchantment(enchantment);
        item.addEnchantment(enchantment, level);
    }
}

File: src/main/kotlin/io/github/narutopig/bors/enchanting/util/Enchanting.kt
package io.github.narutopig.bors.util

import io.github.narutopig.bors.enchanting.CustomEnchants
import io.github.narutopig.bors.enchanting.EnchantmentWrapper
import io.github.narutopig.bors.enchanting.util.EnchantmentData
import io.github.narutopig.bors.util.General.romanToInteger
import io.github.narutopig.bors.util.General.toRoman
import org.bukkit.ChatColor
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemStack

object Enchanting {
    fun calculateLevel(e: Enchantment, a: Int, b: Int): Int {
        /*
        Calculates result of combining the two levels
         */
        val res: Int = if (a != b) a.coerceAtLeast(b) else a + 1
        return res.coerceAtMost(e.maxLevel)
    }

    fun canEnchant(item: ItemStack?, enchantment: Enchantment): Boolean {
        return enchantment.canEnchantItem(item!!) && enchantment.itemTarget.includes(item)
    }

    fun addEnchant(itemStack: ItemStack, enchantment: EnchantmentWrapper, level: Int): ItemStack? {
        val item = itemStack.clone()
        if (!canEnchant(item, enchantment)) return null
        val itemMeta = item.itemMeta ?: return null
        val lore = ArrayList<String>()
        val enchants = getItemCustomEnchants(item)

        enchants[enchantment] = level

        for ((enchant, lvl) in enchants) {
            lore.add("${ChatColor.GRAY}${enchant.name} ${toRoman(lvl)}")
        }

        itemMeta.lore = lore
        item.itemMeta = itemMeta

        return item
    }

    /*
    * @param item Item to check for the enchantment data
    * @param enchantment Name of enchantment as is displayed in Minecraft English (US)
     */
    fun getEnchantmentData(item: ItemStack, enchantment: String): EnchantmentData {
        val itemMeta = item.itemMeta

        if (itemMeta == null) {
            return EnchantmentData(false, 0)
        } else {
            val lore = itemMeta.lore

            return if (lore == null) {
                EnchantmentData(false, 0)
            } else {
                var hasEnchant = false
                var level = 0

                lore.forEach {
                    val stuff = ChatColor.stripColor(it)!!.split(" ")
                    if (stuff[0] == enchantment) hasEnchant = true
                    level = romanToInteger(stuff[1])
                }

                EnchantmentData(hasEnchant, level)
            }
        }
    }

    private fun getItemCustomEnchants(item: ItemStack?): MutableMap<EnchantmentWrapper, Int> {
        val res = mutableMapOf<EnchantmentWrapper, Int>()

        if (item == null) return res
        val itemMeta = item.itemMeta ?: return res
        val lore = itemMeta.lore ?: return res
        if (lore.isEmpty()) return res

        for (line in lore) {
            val l = ChatColor.stripColor(line) ?: continue
            if (l.trim().isEmpty()) continue
            val stuff = l.split(" ")
            var enchant: EnchantmentWrapper
            try {
                enchant = CustomEnchants.getCustomEnchant(stuff[0].trim())
            } catch (e: NoSuchElementException) {
                continue
            }
            val level = romanToInteger(stuff[1].trim())

            if (res.getOrDefault(enchant, 0) < level) {
                res[enchant] = level
            }
        }

        return res
    }
}